compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/TailrecLowering.kt

```diff
@@ -117,69 +128,61 @@ private class BodyTransformer(
+    }
 
-        // For each specified argument set the corresponding variable to it in the correct order:
-        parameterToArgument.forEach { (parameter, argument) ->
-            at(argument)
-            // Note that argument can use values of parameters, so it is important that
-            // references to parameters are mapped using `parameterToNew`, not `parameterToVariable`.
-            +irSet(parameterToVariable[parameter]!!.symbol, argument)
+    private fun IrBuilderWithScope.genTailCall(expression: IrCall) = this.irBlock(expression) {
+        // Get all specified arguments:
+        val parameterToArgument = expression.getArgumentsWithIr().associateTo(mutableMapOf()) { (parameter, argument) ->
+            // Note that we create `val`s for those parameters so that if some default value contains an object
+            // that captures another parameter, it won't capture it as a mutable ref.
+            parameter to irTemporary(argument)
         }
 
-        val specifiedParameters = parameterToArgument.map { (parameter, _) -> parameter }.toSet()
-
         // For each unspecified argument set the corresponding variable to default:
         parameters
-            .filter { it !in specifiedParameters }
+            .filter { it !in parameterToArgument }
             .let { if (properComputationOrderOfTailrecDefaultParameters) it else it.asReversed() }
-            .forEach { parameter ->
-
+            .associateWithTo(parameterToArgument) { parameter ->
                 val originalDefaultValue = parameter.defaultValue?.expression ?: throw Error("no argument specified for $parameter")
-
                 // Copy default value, mapping parameters to variables containing freshly computed arguments:
                 val defaultValue = originalDefaultValue
-                    .deepCopyWithVariables()
-                    .transform(object : IrElementTransformerVoid() {
-
+                    .deepCopyWithVariables().patchDeclarationParents(parent)
+                    .transform(object : VariableRemapper(parameterToArgument) {
                         override fun visitGetValue(expression: IrGetValue): IrExpression {
-                            expression.transformChildrenVoid(this)
-
-                            val variable = parameterToVariable[expression.symbol.owner] ?: return expression
-                            return IrGetValueImpl(
-                                expression.startOffset, expression.endOffset, variable.type,
-                                variable.symbol, expression.origin
-                            )
+                            // If this parameter references a different parameter declared later, produce null:
+                            if (expression.symbol.owner.let { it is IrValueParameter && it.parent == irFunction && it !in parameterToArgument })
+                                return IrConstImpl.defaultValueForType(startOffset, endOffset, expression.type.makeNullable())
```

Comment:
Is that safe enough? Do we have a test for that case?

Reply:
This is consistent with what non-tailrec functions do:
```
fun foo(x: () -> String? = { y }, y: String = "fail"): String? {
    if (y == "start")
        return foo()
    return x()
}

fun main() = println(foo(y = "start")) // null
```
Currently adding `tailrec` changes the output to `start`. I think I saw a non-tailrec version of a test like this.

Reply:
I see, it's weird that non-null function returns `null` in such code
```
fun foo(x: () -> String = { y }, y: String = "fail"): String {
    if (y == "start")
        return foo()
    return x()
}

fun main() = println(foo(y = "start"))
```

Reply:
Turns out replacing variable reads with constants breaks the constructor signature due to changing the capture set, so the copies of the object end up being different...oops.

Reply:
Fixed by adding more temporaries & added a test.

Reply:
>I see, it's weird that the non-null function returns null in such code

FIR reports UNINITIALIZED_PARAMETER here btw.

Reply:
Just in case changing `String?` to `String` doesn't break something else?

Reply:
The temporaries are specifically created with the original type even if it violates null safety, so it shouldn't break anything.